adr/0025-logging.md

```diff
@@ -0,0 +1,235 @@
+| Successf | Info | Seems to be a info log with a checkmark prefix. |
+| Question | ? | Open question how to resolve this. |
+| Questionf | ? | Open question how to resolve this. |
+| Note | Info | Should just be an info or maybe a debug log. |
+| Notef | Info | Should just be an info or maybe a debug log. |
+| Title | ? | Not sure how this should be replaced as it formats with separator. |
+| HeaderInfof | ? | |
+| HorizontalRule | ? | |
+| JSONValue | N/A | Should be replaced with a marshal. |
+| Paragraph | ? | |
+| Paragraphn | ? | |
+| PrintDiff | ? | |
+| Table | ? | Need to come up with a good syntax for functions to return output that can print as a table. |
+| ColorWrap | ? | Should this be used? | 
+| PrintConnectStringTable | N/A | This logic should not exist in the message package. |
+| PrintCredentialTable | N/A | This logic should not exist in the message package. |
+| PrintComponentCredential | N/A | This logic should not exist in the message package. |
+| PrintCredentialUpdates | N/A | This logic should not exist in the message package. |
+| Spinner | Interface | New Spinner interface. |
+| ProgressBar | Interface | New progress bar interface. |
+
+The majority of simpel logging changes should be possible with little signature changes. Replacing the existing output with a slog interface would allow other implementations to be used. A challenge initially may be the changes to table output formatting to make it work properly. This change will require some refactoring of existing code. A requirement for the changes is that they have to improve the UX for users looking at log files. As I understand the present day the spinners will cause a new line everytime they update, resulting in a lot of bloat. A goal should be to make sure that this does not happen in the future.
+
+Spinners and progress bars however are a bit more challenging. They need to be refactored so that they are no long instantiated outside of the CLI code. They should instead be passed as a function parameter to each function that needs them. A good example of a project that solves the problem in a similar manner is Kind. In Kind they create a [status object from the logger](https://github.com/kubernetes-sigs/kind/blob/7799e72306db315ea4f4b1cac90ff68404da4f28/pkg/internal/cli/status.go#L39) which they then [pass to where it is needed](https://github.com/kubernetes-sigs/kind/blob/7799e72306db315ea4f4b1cac90ff68404da4f28/pkg/cluster/internal/create/create.go#L133). Doing so results in a single status object created which is reused where ever it is needed. A lot of inspiration can be take from how Kind delas with CLI output. While they use klog instead of slog there are a lot of similarities. They have for example a check if the output is in a terminal or not, and will disable the spinner accordingly. 
+
+Here is a suggestion for how a thread safe spinner could be implemented with a shared logger. This also allows for paralell spinners and progress bars.
```

Comment:
Have you verified that this is thread-safe? Running `go run -race main.go` shows many data races in the pterm printers, likely related to issue #2558. It seems upstream changes to the pterm printers might be needed for thread safety

Reply:
I initially did not understand all of the race issues in pterm. After looking through things it is obvious that there are issues in pterm that we cannot fix outside of the project. We have two options. Either we wait for pterm to fix the issue internally, or we find and alternative to pterm that offer the same features. It is slightly difficult to find a project that supports multi writer ontop of spinners and progress bars that is not from charm.